vermouth/processors/generate_coordinates.py

```diff
@@ -0,0 +1,382 @@
+        # is defined in a vsn
+        for interaction in molecule.interactions["virtual_sitesn"]:
+            if idx == interaction.atoms[0]:
+                vs_ref_pos = np.zeros((len(interaction.atoms) - 1, 3), dtype=float)
+                for pdx, build_node in enumerate(interaction.atoms[1:]):
+                    vs_ref_pos[pdx, :] = molecule.nodes[build_node]["position"]
+                # if one of the reference atoms is not built we don't
+                # update the molecule
+                vs_pos =  np.average(vs_ref_pos, axis=0)
+                if not all(np.isnan(vs_pos)):
+                    molecule.nodes[idx]["position"] = vs_pos
+                break
+    return molecule
+
+
+class GenerateCoordinates(Processor):
+    """
+    A processor for generating coordinates for atoms without.
+    """
+    def __init__(self, contact=0.5):
+        super().__init__()
+        self.contact = contact
+        self.coordinates = None
+
+    def run_system(self, system):
+        """
+        Process `system`.
+
+        Parameters
+        ----------
+        system: vermouth.system.System
+            The system to process. Is modified in-place.
+        """
+
+        self.coordinates = np.stack(
+            [
+                a.get('position', (np.nan, np.nan, np.nan))
+                for m in system.molecules
+                for i, a in m.atoms
+            ]
+        )
+
+        for molecule in system.molecules:
+            self.run_molecule(molecule)
+
+    def run_molecule(self, molecule):
+        """
+        Process a single molecule. Must be implemented by subclasses.
+
+        Parameters
+        ----------
+        molecule: vermouth.molecule.Molecule
+            The molecule to process.
+
+        Returns
+        -------
+        vermouth.molecule.Molecule
+            The provided molecule with complete coordinates
+        """
+        # Missing atoms - those without 'positions'
+        missing = get_missing_atoms(molecule)
+        prev = len(molecule)
+        while prev - len(missing) > 0:
+            # Limbs: subgraphs of missing atoms with anchors and bases
+            limbs = []
+            for g in nx.connected_components(molecule.subgraph(missing)):
+                anchors = list(get_anchors(molecule, g))
+                bases = [ list(get_anchors(molecule, [a])) for a in anchors ]
+                limbs.append((anchors, bases, g))
+            limbs.sort()
+
+            # Segment: subgraphs of limbs sharing anchors and bases
+            segments = []
+            prev = None
+            for lim in limbs:
+                if lim[0] != prev:
+                    segments.append([])
+                    prev = lim[0]
+                segments[-1].append(lim)
+            segments = [ Segment(molecule, limbs) for limbs in segments ]
+
+            # Extrude segments one layer at a time
+            while segments:
+                s = segments.pop(0)
```

Comment:
Nit: Not ideal to be popping from the left all the time. Is it possible to reverse sort the limbs above, and pop from the right here?